text_volpiano_alignment.py

```diff
@@ -0,0 +1,196 @@
+"""
+Module that aligns the syllabified text and melody (encoded in volpiano) of a chant. This
+module assumes that text and melody have been entered according to the conventions of Cantus
+Database (more details about these conventions can be found 
+at https://cantusdatabase.org/documents). See the README for more information.
+
+Use align_text_and_volpiano to align the text and melody strings of a chant.
+"""
+import re
+import logging
+from itertools import zip_longest, takewhile
+from cantus_text_syllabification import syllabify_text
+
+# A string containing all valid volpiano characters in Cantus
+# Database. Used to check for invalid characters.
+VALID_VOLPIANO_CHARS = "-19abcdefghijklmnopqrsyz)ABCDEFGHIJKLMNOPQRSYZ3467]"
+
+
+def _preprocess_volpiano(volpiano_str: str) -> str:
+    """
+    Prepares volpiano string for alignment with text:
+    - Checks for any invalid characters
+    - Ensures proper spacing around barlines ("3" and "4") and missing
+        music indicators ("6")
+    - Ensures volpiano string has an ending barline ("3" or "4")
+
+    volpiano_str [str]: volpiano string
+
+    returns [str]: preprocessed volpiano string
+    """
+    processed_str = ""
+    volpiano_str_len = len(volpiano_str)
+    for i, char in enumerate(volpiano_str):
+        # Check if char is valid
+        if char not in VALID_VOLPIANO_CHARS:
+            logging.debug("Removed invalid character (%s) in volpiano string.", char)
+            continue
+        # Check if char is a barline or missing music indicator and ensure
+        # proper spacing
+        if (char in "346") and (i != volpiano_str_len - 1):
+            # Add proper spacing before barline
+            while processed_str[-3:] != "---":
+                processed_str += "-"
+            # Add barline character
+            processed_str += char
+            # Add proper spacing after barline
+            num_hyph_ahead = sum(
```

Comment:
not `num_hyph_behind`? This calculates the number of hyphens following (i.e. behind) the barline, doesn't it?

Reply:
I think I will say `num_hyphen_next`. To me, if I am thinking about iterating through the string, the next characters are those that are "ahead" and the previous characters are "behind"...but I see these terms are v ambiguous, where as "next" and "previous" seem clear?